src/mgxs_interface.cpp

```diff
@@ -0,0 +1,237 @@
+#include "mgxs_interface.h"
+
+namespace openmc {
+
+//==============================================================================
+// Mgxs data loading interface methods
+//==============================================================================
+
+void
+add_mgxs_c(hid_t file_id, char* name, const int energy_groups,
+     const int delayed_groups, const int n_temps, double temps[], int& method,
+     const double tolerance, const int max_order,
+     const bool legendre_to_tabular, const int legendre_to_tabular_points,
+     const int n_threads)
+{
+  //!! mgxs_data.F90 will be modified to just create the list of names
+  //!! in the order needed
+  // Convert temps to a vector for the from_hdf5 function
+  double_1dvec temperature;
+  temperature.assign(temps, temps + n_temps);
+
+  write_message("Loading " + std::string(name) + " data...", 6);
+
+  // Check to make sure cross section set exists in the library
+  hid_t xs_grp;
+  if (object_exists(file_id, name)) {
+    xs_grp = open_group(file_id, name);
+  } else {
+    fatal_error("Data for " + std::string(name) + " does not exist in "
+                + "provided MGXS Library");
+  }
+
+  Mgxs mg;
+  mg.from_hdf5(xs_grp, energy_groups, delayed_groups,
+       temperature, method, tolerance, max_order, legendre_to_tabular,
+       legendre_to_tabular_points, n_threads);
+
+  nuclides_MG.push_back(mg);
+}
+
+//==============================================================================
+
+bool
+query_fissionable_c(const int n_nuclides, const int i_nuclides[])
+{
+  bool result = false;
+  for (int n = 0; n < n_nuclides; n++) {
+    if (nuclides_MG[i_nuclides[n] - 1].fissionable) result = true;
+  }
+  return result;
+}
+
+//==============================================================================
+
+void
+create_macro_xs_c(char* mat_name, const int n_nuclides,
+     const int i_nuclides[], const int n_temps, const double temps[],
+     const double atom_densities[], int& method, const double tolerance,
+     const int n_threads)
+{
+  Mgxs macro;
+  if (n_temps > 0) {
+    // // Convert temps to a vector
+    double_1dvec temperature;
+    temperature.assign(temps, temps + n_temps);
+
+    // Convert atom_densities to a vector
+    double_1dvec atom_densities_vec;
+    atom_densities_vec.assign(atom_densities, atom_densities + n_nuclides);
+
+    // Build array of pointers to nuclides_MG's Mgxs objects needed for this
+    // material
+    std::vector<Mgxs*> mgxs_ptr(n_nuclides);
+    for (int n = 0; n < n_nuclides; n++) {
+      mgxs_ptr[n] = &nuclides_MG[i_nuclides[n] - 1];
+    }
+
+    macro.build_macro(mat_name, temperature, mgxs_ptr, atom_densities_vec,
+         method, tolerance, n_threads);
```

Comment:
This also looks like it could be a constructor.

Reply:
- [x]